### DISCUSSION TITLE
Options for python code completion/hinting and navigation

### DISCUSSION THREAD

**Discussion by aspitarl** [IDAES/idaes-pse#1547] [Q&A] [answered]
I am developing with IDAES in VS Code. It appears that there is limited ability to use features such as intellisense or navigate to class definitions easily. This seems to come from two main things: 

1. The use of decorators (`@declare_process_block_class`) to define e.g. `Compressor` from `CompressorData`. So the language server does not recognize `Compressor` and it is not easy to go to the class definition. It's not clear to me why the standard python initialization and class inheritance isn't sufficient, but I assume there's a good reason. 
2. The use of CONFIG objects instead of standard key word arguments in an __init__ function. I see that this comes from pyomo, so that makes sense to keep things consistent. 

So far, I've been resorting to searching through the source code to find how something works. Is there any method for easier development in an IDE that I am missing? For example, the CONFIG objects have doc strings, but it appears this is mainly for the online documentation. 

Any information on best practices for development environment would be helpful. Thanks.

**Comment by adowling2** [comment]
Thank you for posting on the Discussion! One thing I need to do when using VS Code (including packages besides IDAES) is ensure I am loading the correct conda environment.

In addition to looking at the source code, I find the document (readthedocs) and examples helpful.

**Reply by aspitarl** [reply]
Thanks, yes I have the correct environment activated. For example  `from idaes.models.unit_models.pressure_changer import CompressorData` goes to the correct definition. But it's a good note to include here as you can still get code completion with utilities like model statistics/Diagnostics etc. since those are normal python functions.

**Comment by dallan-keylogic** [comment]
As you noticed, (2) is a design pattern inherited from Pyomo. The advantage is that most input validation is automated and we can clearly show users which options are allowed for a field.  (1) is another design pattern inherited from Pyomo to automate the generation of the "diamond" class structure that allows us to create indexed and scalar varieties of the same class with a single declaration.

Unfortunately, I don't have a ready made solution. Hopefully someone from Pyomo, like @blnicho  or @jsiirola can answer your question.

**Comment by lbianchi-lbl** [comment, marked as answer]
Thanks a lot @aspitarl for bringing this up.

- I agree with your initial assessment that those non-standard implementations are indeed the limiting factor for anything that relies on static analysis
- As @dallan-keylogic confirmed, there are reasons why these solutions are in place. A while ago I explored a few alternatives API to replace  `@declare_process_block_class`, but due to backward compatibility and limited available resources, it is unlikely that the API will change to a more static-analysis-friendly one anytime soon
- However, there might be good news on the horizon since I've started investigating using type stub files (`.pyi`) that can be used to supply information to IDEs without changing the runtime code. You can follow progress on that front in #1548

**Reply by aspitarl** [reply]
Stub files sounds like the way to go. Thanks for sharing, I actually have been looking for something like that. I'll continue the discussion in the issue.
